lib/sax.js

```diff
@@ -538,86 +538,228 @@ function _copy(source, target) {
+			if (cdata) {
 				domBuilder.startCDATA();
-				domBuilder.characters(source, start + 9, end - start - 9);
+				domBuilder.characters(cdata, 9, cdata.length - 12);
 				domBuilder.endCDATA();
-				return end + 3;
+				return index;
+			} else {
+				return errorHandler.fatalError('Invalid CDATA starting at position ' + start);
 			}
-			//<!DOCTYPE
-			//startDTD(java.lang.String name, java.lang.String publicId, java.lang.String systemId)
-			var matchs = split(source, start);
-			var len = matchs.length;
-			if (len > 1 && /!doctype/i.test(matchs[0][0])) {
-				var name = matchs[1][0];
-				var pubid = false;
-				var sysid = false;
-				if (len > 3) {
-					if (/^public$/i.test(matchs[2][0])) {
-						pubid = matchs[3][0];
-						sysid = len > 4 && matchs[4][0];
-					} else if (/^system$/i.test(matchs[2][0])) {
-						sysid = matchs[3][0];
-					}
+		case 'D': {
+			// should be DOCTYPE
+			var doctype = {
+				name: undefined,
+				publicId: undefined,
+				systemId: undefined,
+				internalSubset: undefined,
+			};
+
+			if (!compare('<!DOCTYPE')) {
+				return errorHandler.fatalError('Expected <!DOCTYPE');
+			}
+			skip(9);
+			if (skipBlanks() < 1) {
+				return errorHandler.fatalError('No blanks between DOCTYPE and document name');
+			}
+
+			// Parse the DOCTYPE name
+			doctype.name = getMatch(grammar.Name);
+			if (!doctype.name) return errorHandler.fatalError('DOCTYPE without name');
+			skipBlanks();
+
+			// Check for External ID and SystemId
+			if (compare('PUBLIC')) {
+				skip(6);
+				if (skipBlanks() < 1) {
+					return errorHandler.fatalError('No blanks between PUBLIC and public identifier');
 				}
-				var lastMatch = matchs[len - 1];
-				domBuilder.startDTD(name, pubid, sysid);
-				domBuilder.endDTD();
+				doctype.publicId = getMatch(grammar.PubidLiteral);
+				if (!doctype.publicId) return errorHandler.fatalError('PUBLIC without public identifier');
+				if (skipBlanks() < 1) {
+					return errorHandler.fatalError('No blanks between public identifier and SystemId');
+				}
+				doctype.systemId = getMatch(grammar.SystemLiteral);
+				if (!doctype.systemId) return errorHandler.fatalError('PUBLIC without SystemId');
+			} else if (compare('SYSTEM')) {
+				skip(6);
+				if (skipBlanks() < 1) {
+					return errorHandler.fatalError('No blanks between SYSTEM and SystemId');
+				}
+				doctype.systemId = getMatch(grammar.SystemLiteral);
+				if (!doctype.systemId) return errorHandler.fatalError('SYSTEM without SystemId');
+			}
+
+			// Parse internal subset
+			skipBlanks();
 
-				return lastMatch.index + lastMatch[0].length;
+			if (char() === '[') {
+				skip(1);
+				var intSubsetStart = index;
+				skipBlanks();
+				while (index < source.length) {
+					var current = null;
+					// Only in external subset
+					// if (char() === '<' && char(1) === '!' && char(2) === '[') {
+					// 	parseConditionalSections();
+					// } else
+					if (char() === '<' && char(1) === '!') {
```

Comment:
If this is a common case the compare method could also have special handling for text with length 2, avoiding the creation of a substring

Reply:
This is less expensive than having an additional check. I would consider using the compare function for >3 chars. and this individual `char(n)` approach for anything less.

Reply:
Im not proposing to remove the `char` function just that doing a loop in `compare` aka `substringStartsWith` like we do on `skipBlanks` instead of doing the substring and === might make the code easier to understand and maintain and faster. And if we know that there is a limit of e.g. 20 where the loop is slower we could decide based on the length of the string. But I'm not sure we ever pass such long strings to that function, so maybe that check is not needed.
I guess we should also benchmark that.

Reply:
I have benchmarked your proposal:

```
function compare(text, index) {
  return source.substring(index, index + text.length) === text;
}

function compareNew(text, index) {
  for (let i = 0; i < text.length; i++) {
    if (source.charAt(index + i) !== text[i]) {
      return false;
    }
  }
  return true;
}
```

```
Original compare - size 1: 7.834ms
New compare - size 1: 7.179ms

Original compare - size 2: 16.994ms
New compare - size 2: 5.27ms

Original compare - size 3: 9.294ms
New compare - size 3: 7.11ms

Original compare - size 4: 9.846ms
New compare - size 4: 9.396ms
```

From there on original compare starts gaining significant advantage.

It looks like this approach would only be noticeably faster for a string of length 2 and marginally faster for string of length 3.

Given this, I would prefer to leave the compare, (or now `substringStartsWith`) as is.